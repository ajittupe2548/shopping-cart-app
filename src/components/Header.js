import React, { memo, useCallback } from "react";
import { useSelector, useDispatch } from "react-redux";
import { toggleCart } from "../store/cartSlice";
import styles from "./Header.module.scss";

const Header = memo(() => {
  const dispatch = useDispatch();
  const totalQuantity = useSelector(state => state.cart.totalQuantity);

  const handleCartClick = useCallback(() => {
    dispatch(toggleCart());
  }, [dispatch]);

  return (
    <header className={styles.header}>
      <div className={styles.headerContent}>
        <h1 className={styles.logo}>Shopping Cart</h1>
        <button className={styles.cartIcon} onClick={handleCartClick}>
          ðŸ›’ Cart
          {totalQuantity > 0 && (
            <span className={styles.cartCount}>{totalQuantity}</span>
          )}
        </button>
      </div>
    </header>
  );
});

Header.displayName = "Header";

export default Header;
