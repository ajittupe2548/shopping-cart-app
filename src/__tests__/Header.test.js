import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import Header from "../components/Header";
import cartReducer from "../store/cartSlice";

const renderWithRedux = (component, initialState = {}) => {
  const store = configureStore({
    reducer: {
      cart: cartReducer,
    },
    preloadedState: initialState,
  });

  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
};

describe("Header", () => {
  test("renders header with title and cart button", () => {
    renderWithRedux(<Header />);

    expect(screen.getByText("Shopping Cart")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /cart/i })).toBeInTheDocument();
  });

  test("does not show cart count when cart is empty", () => {
    renderWithRedux(<Header />);

    expect(screen.queryByText(/\d+/)).not.toBeInTheDocument();
  });

  test("shows correct cart count when items are in cart", () => {
    const initialState = {
      cart: {
        items: [
          { id: 1, quantity: 2 },
          { id: 2, quantity: 1 },
        ],
        totalQuantity: 3,
        totalAmount: 59.98,
        isOpen: false,
      },
    };

    renderWithRedux(<Header />, initialState);

    expect(screen.getByText("3")).toBeInTheDocument();
  });

  test("toggles cart when cart button is clicked", () => {
    const { store } = renderWithRedux(<Header />);

    const cartButton = screen.getByRole("button", { name: /cart/i });
    fireEvent.click(cartButton);

    const state = store.getState();
    expect(state.cart.isOpen).toBe(true);
  });

  test("matches snapshot with empty cart", () => {
    const initialState = {
      cart: {
        items: [],
        totalQuantity: 0,
        totalAmount: 0,
        isOpen: false,
      },
    };

    const { container } = renderWithRedux(<Header />, initialState);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("matches snapshot with items in cart", () => {
    const initialState = {
      cart: {
        items: [{ id: 1, quantity: 3 }],
        totalQuantity: 5,
        totalAmount: 100,
        isOpen: false,
      },
    };

    const { container } = renderWithRedux(<Header />, initialState);
    expect(container.firstChild).toMatchSnapshot();
  });
});
