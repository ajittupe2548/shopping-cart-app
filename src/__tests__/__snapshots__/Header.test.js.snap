// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Header matches snapshot with empty cart 1`] = `
<header
  class="header"
>
  <div
    class="headerContent"
  >
    <h1
      class="logo"
    >
      Shopping Cart
    </h1>
    <button
      class="cartIcon"
    >
      ðŸ›’ Cart
    </button>
  </div>
</header>
`;

exports[`Header matches snapshot with items in cart 1`] = `
<header
  class="header"
>
  <div
    class="headerContent"
  >
    <h1
      class="logo"
    >
      Shopping Cart
    </h1>
    <button
      class="cartIcon"
    >
      ðŸ›’ Cart
      <span
        class="cartCount"
      >
        5
      </span>
    </button>
  </div>
</header>
`;
